packages/remix-react/components.tsx

```diff
@@ -929,16 +1171,177 @@ export function useFetcher<TData = any>(): FetcherWithComponents<
   SerializeFrom<TData>
 > {
   let fetcherRR = useFetcherRR();
-  let { Form, submit, load } = fetcherRR;
-
-  // TODO: For SSR we can just return IDLE_FETCHER for now, we'll handle
-  // this as part of hydration work
+  let remixFetcher = convertRouterFetcherToRemixFetcher({
+    state: fetcherRR.state,
+    data: fetcherRR.data,
+    formMethod: fetcherRR.formMethod,
+    formAction: fetcherRR.formAction,
+    formData: fetcherRR.formData,
+    formEncType: fetcherRR.formEncType,
+    " _hasFetcherDoneAnything ": fetcherRR[" _hasFetcherDoneAnything "],
+  });
   return {
-    Form,
-    submit,
-    load,
-    ...IDLE_FETCHER,
+    ...remixFetcher,
+    load: fetcherRR.load,
+    submit: fetcherRR.submit,
+    Form: fetcherRR.Form,
+  };
+}
+
+function convertRouterFetcherToRemixFetcher(
+  fetcherRR: Omit<ReturnType<typeof useFetcherRR>, "load" | "submit" | "Form">
+): Fetcher {
+  let { state, formMethod, formAction, formEncType, formData, data } =
+    fetcherRR;
+
+  let isActionSubmission =
+    formMethod != null &&
+    ["POST", "PUT", "PATCH", "DELETE"].includes(formMethod.toUpperCase());
+
+  if (state === "idle") {
+    if (fetcherRR[" _hasFetcherDoneAnything "] === true) {
+      let fetcher: FetcherStates["Done"] = {
+        state: "idle",
+        type: "done",
+        submission: undefined,
+        data,
+      };
+      return fetcher;
+    } else {
+      let fetcher: FetcherStates["Idle"] = IDLE_FETCHER;
+      return fetcher;
+    }
+  }
+
+  if (
+    state === "submitting" &&
+    formMethod &&
+    formAction &&
+    formEncType &&
+    formData
+  ) {
+    if (isActionSubmission) {
+      // Actively submitting to an action
+      let fetcher: FetcherStates["SubmittingAction"] = {
+        state,
+        type: "actionSubmission",
+        submission: {
+          method: formMethod.toUpperCase() as ActionSubmission["method"],
+          action: formAction,
+          encType: formEncType,
+          formData: formData,
+          // TODO: this is created as a random hash value in useSubmitImpl in
+          // Remix today. We do not have this key in react router as we
+          // flattened submissions down onto the fetcher.  We can't recreate in
+          // this back-compat layer in a stable manner for useFetchers because
+          // we don't have a stable fetcher identity.  So we could:
+          //  - Expose the fetcher key from the router (might make sense if
+          //    we're considering adding useFetcher({ key }) anyway
+          //  - Expose a hidden field with a stable identifier on the fetcher
+          //    like we did for _hasFetcherDoneAnything
+          key: "todo-not-implemented-yet",
```

Comment:
Router doesn't have submission keys - need to think about the best way to tackle this.

Maybe we just add a navigation key to router?

```js
navigation = {
  state,
  location,  // has a key
  form*,
  key, // populate on _all_ navigations, submisison or not
       // does not change on redirects during a naviagtion
}
```

Reply:
Might be able to solve this with `location.state._submissionKey` which would allow us to track this key across submission redirects